recipes/CoVoST/AST/train_w2v2_llama.py

```diff
@@ -0,0 +1,550 @@
+#!/usr/bin/env python3
+"""Recipe for training an XLSR plus LLaMA 3 speech translation system
+on CoVoST. The system employs an XLSR encoder and a LLaMA3 decoder. A simple
+project concatenating frames is trained between XLSR and LLaMA.
+
+To run this recipe, do the following:
+> python train_xlsr_llama.py hparams/xlsr_llama3.yaml
+
+
+Authors
+ * Titouan Parcollet 2025
+"""
+import sys
+
+import torch
+import torchaudio
+import transformers
+from hyperpyyaml import load_hyperpyyaml
+
+import speechbrain as sb
+from speechbrain.dataio.dataio import length_to_mask
+from speechbrain.utils.distributed import run_on_main
+from speechbrain.utils.logger import get_logger
+
+logger = get_logger(__name__)
+
+
+# Define training procedure
+class AST(sb.core.Brain):
+    def compute_forward(self, batch, stage):
+        """Forward computations from the waveform batches to the output probabilities."""
+        batch = batch.to(self.device)
+        wavs, wav_lens = batch.sig
+        wavs, wav_lens = wavs.to(self.device), wav_lens.to(self.device)
+        tokens_prompt_translation, tokens_prompt_translation_len = (
+            batch.tokens_prompt_translation
+        )  # Includes prompt and translation
+        prompt_len = batch.prompt_len
+
+        # Turn padding in the speech to zero. We need to do this in case of leaks, because LLAMA padding is using int of value 120k+ which will corrupt A LOT the signal in case of leak.
+        audio_len = wavs.shape[1]
+        abs_len = torch.round(wav_lens * audio_len)
+        audio_attn_mask = length_to_mask(abs_len)
+        wavs = wavs * audio_attn_mask
+
+        # Add waveform augmentation if specified.
+        if (
+            stage == sb.Stage.TRAIN
+            and hasattr(self.hparams, "wav_augment")
+            and self.optimizer_step > self.hparams.augment_warmup
+        ):
+            wavs, wav_lens = self.hparams.wav_augment(wavs, wav_lens)
+
+        wavs = self.hparams.normalize(wavs, wav_lens)
+
+        # Forward Speech Modules
+        feats = self.modules.wav2vec2(wavs, wav_lens)
+        down_feats = self.modules.feat_downsampler(feats)
+        down_feats_proj = self.modules.proj(down_feats)
+
+        # Format input for LLM; [ audio emb ] + [ prompt emb ] + [ translation emb ]
+        # First get relevant lengths
+        audio_len = down_feats_proj.shape[1]
+        text_len = tokens_prompt_translation.shape[1]
+        audio_prompt_len = (audio_len + prompt_len)[0]
+
+        # Then for the full embedding prompt sequence
+        if hasattr(self.modules.lora_llm, "module"):
+            embeddings = (
+                self.modules.lora_llm.module.model.get_input_embeddings()
+            )
+        else:
+            embeddings = self.modules.lora_llm.model.get_input_embeddings()
```

Comment:
this makes me believe we really need to have a pointer to the raw parameters (i.e. before wrapping the module to DDP). This way, we can just call specific methods that don't need forward passes.

Reply:
this is cost-free to do. We just need to add a single line in core.py that does `self.unwrapped_modules = self.module` before wrapping it in DDP. And this way we can then do `self.unwrapped_modules.lora_llm.get_input_embeddings()` without having to check for `.module.`

Reply:
wondering if maybe instead of calling this at each step, you could just save the embedding in the brain class e.g. `brain.txt_embeddings = get_input_embeddings()` so that in the code you can just call it like this `brain.txt_embeddings(...)` (similar to `brain.tokenizer`).

Reply:
wait. does this code works if we are not using LORA? Here you are making the assumption that the lora block of code is not commented